feat: add a scroll to top btn

12rambau opened the pull request:
Fix #1126 

I added a scroll to top btn to the main article page. It's located 2 * header sidebar from the top and centered on the page. Following instrcution from @choldgraf and the bootstrap implementation (https://mdbootstrap.com/snippets/standard/mdbootstrap/2964350#js-tab-view) I manage to get something lightweight and working. 

The trigger is simply: "when scolling up and under the header initial bottom". I decided to avoid the hidden pixel and to rely on the header bottom to avoid issue when there is a disclaimer. 

let me know what you think. pinging @trallard as you mentioned some a11y issues that I may failed to apply.

trallard commented:
Ooops just noticed this is waiting for me - I have added it to my backlog @12rambau apologies for the delay

12rambau commented:
no problems we all have long dashboards, I'm pretty sure i'm late on many other.

trallard commented on src/pydata_sphinx_theme/assets/styles/base/_base.scss:
```diff
@@ -177,3 +177,16 @@ pre {
+  position: fixed;
+  display: none;
+  top: calc(var(--pst-header-height) + 1rem);
+  left: 50%;
```
Is there a reason why you are using `left: 50%;` vs `left: auto;` is this to avoid clash with other page components?

12rambau replied:
no it's because i'm not expert in css. I assume auto is better ?

12rambau replied:
I just tried what you suggested and the button ended on the far left side of my screen, rolling back to previous version.

trallard commented on src/pydata_sphinx_theme/theme/pydata_sphinx_theme/layout.html:
```diff
@@ -52,6 +52,15 @@
+  {# A tiny helper pixel to detect if we've scrolled #}
+  <div id="pst-scroll-pixel-helper"></div>
+
+  {# the scroll to tpo button #}
```
```suggestion
  {# the scroll to top button #}
```

trallard commented on src/pydata_sphinx_theme/theme/pydata_sphinx_theme/layout.html:
```diff
@@ -52,6 +52,15 @@
+  {# the scroll to tpo button #}
+  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
+    <i class="fa-solid fa-arrow-up"></i>
+    {{ _("Scroll to top") }}
```
I think a less ambiguous term would be helpful for example:
```suggestion
    {{ _("Back to top") }}
```

Since this is meant to be a focusable and keyboard operable element the actual action is to go back to the top rather than a mouse wheel scroll equivalent

trallard reviewed (approved):
Thanks, @12rambau. Since you already use a button element for this, I think it should already have all the correct ARIA added.

I left a couple of comments - nothing blocking.

Also since I could not see the button in the rendered PR is this meant to be enabled by default or would this require user-configuration?

choldgraf reviewed (approved):
This looks great to me - once @trallard is happy I'm happy to merge 👍

12rambau commented:
> Also since I could not see the button in the rendered PR is this meant to be enabled by default or would this require user-configuration?

I checked [this page](https://pydata-sphinx-theme--1332.org.readthedocs.build/en/1332/community/setup.html) and it worked by default as expected. Note that the button only appears if you start scrolling up.

trallard commented:
aha! It took me a while to notice it since it appears right in the middle of the screen and the colour is slightly off the background. 

![Get_started_with_development_—_PyData_Theme_0_13_4dev0_documentation](https://github.com/pydata/pydata-sphinx-theme/assets/23552331/226f86d5-2f57-463b-ad34-60efd1e94dc7)

So I will say that the current positioning + style is not distinct enough from the surrounding context.

I think a better positioning would be placing this at the bottom of the page with a more noticeable styling. See, for example, this website https://ashleemboyer.com/blog/accessible-smooth-scroll-to-top-buttons-with-little-code (I know it is a single-column content page) but the style does make it much easier to spot.

![Accessible__Smooth_Scroll-to-top_Buttons_with_Little_Code___Ashlee_M_Boyer](https://github.com/pydata/pydata-sphinx-theme/assets/23552331/11673737-6979-43fb-860a-75126347eb7b)

Another example is the [gov.uk design system page](https://design-system.service.gov.uk/components/back-link/) where the `Back to top` component is also at the bottom - on their sidebar, but since there is little clutter there it is easy to see it


Tagging @smeragoel in case she has any insights to share with us from an UX perspective

choldgraf commented:
I'm +1 on putting it on the bottom. I'm assuming most people will use "up to top" on mobile and if that assumption is right, their thumb is easier to reach the bottom than the top

trallard commented:
Absolutely! I completely forgot about the thumb safe areas on mobile.

See for ref https://www.smashingmagazine.com/2016/09/the-thumb-zone-designing-for-mobile-users/

12rambau commented:
I've moved the button in the last 20% of screen so it's closer to where we expect to find it. I kept some margin in case people get an big footer. I also changed its color to primary. @trallard let me know if it's more visible to you.

smeragoel commented:
I think having it on the bottom is definitely an improvement from having it on top. However, I don't like the fact that the button overlaps existing content, that is always messy from a UX pov. I quite like the [gov.uk design system page](https://design-system.service.gov.uk/components/back-link/) suggestion made my @trallard. 

I think we can modify the button to be on the bottom right of the screen, since the right sidebar is pretty much empty towards the bottom. That way the button will be visible, and we avoid any sort of overlapping. 

[This article](https://www.nngroup.com/articles/back-to-top/) also outlines some guidelines for `scroll to top` buttons, and they suggest that the [bottom right side](https://www.nngroup.com/articles/back-to-top/#:~:text=Place%20a%20persistent%20Back%20to%20Top%20button%20in%20the%20lower%20right%20side%20of%20the%20page.) of the page is the most intuitive.

12rambau commented:
thanks @smeragoel for your feedback. As this button needs to be available for small devices how can I add it without overlapping the text ? should I hard code an horizontal position or maintain it in the content area for any devices ?

drammock commented:
I would guess it's related, and it's because those other functions are passing args, not kwargs, and you inserted your new kwarg before `n_fft` in `_mt_spectra`. Put your new kwarg at the end.

choldgraf commented:
This looks great. One question: for me the button has black text on blue background and it's a bit hard to read. Is it WCAG friendly?

![Screenshot_20230701-075321](https://github.com/pydata/pydata-sphinx-theme/assets/1839645/03ac38a0-73ae-4b88-8f2b-04e6869a1071)

trallard commented:
Some thoughts

1. I am still not super convinced of the position - but that could be fixed later on also based on feedback
2. The button is in the primary colour, right now there are a lot of components and links with the same colour so that button would be failing contrast requirements when overlapping with the links behind it (they are the same colour so there is a 1:1 contrast)

Suggestions: 

- This PR is still using the old colours so it would be best to merge main here to bring in line with the recent changes 
- Change the colour to our secondary colour (that way it should have enough contrast against the background when overlapping over links and such)
- ensure the text inside the button by using the  `--pst-color-secondary-text` variable that is generated based on WCAG contrast of 4.5:1

choldgraf commented:
Ah good point about colors. In that case I think it's better we merge this and get feedback from people rather than getting things perfect now. I'm +1 to merge as is

trallard commented:
I'd say we should at least fix the text colour inside the button - right now it's using the `--pst-color-text-base` which is inaccessible on both themes (and makes this button different from all the other buttons now on the theme)

![Screenshot_20230701-084041.png](https://github.com/pydata/pydata-sphinx-theme/assets/23552331/0bc5a1da-4631-4d8e-9cb9-e327dee393fe)

choldgraf commented:
As you like, I'll hold off on merging if you prefer to do that fix here instead of a follow up

trallard commented:
Right let's merge and do a quick follow-up PR to fix the colours.

choldgraf commented:
@smeragoel do you wish to hold off on merging or are you ok to merge and iterate?

12rambau commented:
Let me update the branch with the modification suggested and I'll merge it

smeragoel commented:
+1 from me as well to merge it now, and then focus on colours and later positioning!

12rambau commented:
Thanks all for your comments, I've done the color updates and merged. Let's iterate from it afterward
